plugins/core/so11y.go

```diff
@@ -0,0 +1,165 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package core
+
+import (
+	"sync"
+	"time"
+
+	"github.com/apache/skywalking-go/plugins/core/metrics"
+)
+
+var (
+	instance *So11y
+	once     sync.Once
+)
+
+type So11y struct {
+	propagatedContextCounter       metrics.Counter
+	propagatedIgnoreContextCounter metrics.Counter
+
+	samplerContextCounter       metrics.Counter
+	samplerIgnoreContextCounter metrics.Counter
+
+	finishContextCounter       metrics.Counter
+	finishIgnoreContextCounter metrics.Counter
+
+	leakedContextCounter       metrics.Counter
+	leakedIgnoreContextCounter metrics.Counter
+
+	errorCounterMap     sync.Map
+	interceptorTimeCost metrics.Histogram
+}
+
+func GetSo11y() *So11y {
+	once.Do(func() {
+		instance = &So11y{}
+	})
+	return instance
+}
+
+func (s *So11y) MeasureTracingContextCreation(t *Tracer, isForceSample, isIgnored bool) {
+	if isForceSample {
+		if isIgnored {
+			if s.propagatedIgnoreContextCounter == nil {
```

Comment:
Should we instance all metrics when creating the So11y structure, I think all metrics will created multiple times when running in multiple go-routines.